File: BraintreeCore/src/main/java/com/braintreepayments/api/AuthorizationLoader.java
package com.braintreepayments.api;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

class AuthorizationLoader {

    private Authorization authorization;
    private final ClientTokenProvider clientTokenProvider;

    AuthorizationLoader(@Nullable String initialAuthString, @Nullable ClientTokenProvider clientTokenProvider) {
        this.clientTokenProvider = clientTokenProvider;
        if (initialAuthString != null) {
            this.authorization = Authorization.fromString(initialAuthString);
        }
    }

    void loadAuthorization(@NonNull final AuthorizationCallback callback) {
        if (authorization != null) {
            callback.onAuthorizationResult(authorization, null);
        } else if (clientTokenProvider != null) {
            clientTokenProvider.getClientToken(new ClientTokenCallback() {
                @Override
                public void onSuccess(@NonNull String clientToken) {
                    authorization = Authorization.fromString(clientToken);
                    callback.onAuthorizationResult(authorization, null);
                }

                @Override
                public void onFailure(@NonNull Exception error) {
                    callback.onAuthorizationResult(null, error);
                }
            });
        } else {
            String clientSDKSetupURL
                = "https://developer.paypal.com/braintree/docs/guides/client-sdk/setup/android/v4#initialization";
            String message = String.format("Authorization required. See %s for more info.", clientSDKSetupURL);
            callback.onAuthorizationResult(null, new BraintreeException(message));
        }
    }

    @Nullable
    Authorization getAuthorizationFromCache() {
        return authorization;
    }
}

File: BraintreeCore/src/main/java/com/braintreepayments/api/AuthorizationLoader.kt
package com.braintreepayments.api

internal class AuthorizationLoader(
    initialAuthString: String?,
    private val clientTokenProvider: ClientTokenProvider?
) {
    // cache initial auth if available
    var authorizationFromCache = initialAuthString?.let { Authorization.fromString(it) }

    fun loadAuthorization(callback: AuthorizationCallback) {
        if (authorizationFromCache != null) {
            callback.onAuthorizationResult(authorizationFromCache, null)
        } else if (clientTokenProvider != null) {
            clientTokenProvider.getClientToken(object : ClientTokenCallback {
                override fun onSuccess(clientToken: String) {
                    authorizationFromCache = Authorization.fromString(clientToken)
                    callback.onAuthorizationResult(authorizationFromCache, null)
                }

                override fun onFailure(error: Exception) {
                    callback.onAuthorizationResult(null, error)
                }
            })
        } else {
            val clientSDKSetupURL =
                "https://developer.paypal.com/braintree/docs/guides/client-sdk/setup/android/v4#initialization"
            val message = "Authorization required. See $clientSDKSetupURL for more info."
            callback.onAuthorizationResult(null, BraintreeException(message))
        }
    }

    fun invalidateClientToken() {
        // only invalidate client token cache if we can fetch a new one with a client token provider
        if (clientTokenProvider != null) {
            authorizationFromCache = null
        }
    }
}
